### PR TITLE
Mixed precision monin_obukhov

### PR DESCRIPTION
**Pull request by mcallic2**
**Description**
Adds mixed precision to monin_obukhov

Fixes # (issue)

**How Has This Been Tested?**
Autotools using oneapi compiler

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

File: monin_obukhov/include/monin_obukhov.inc [hunk 1]
```diff
@@ -252,153 +94,174 @@ if(present(avail)) then
...
-real, intent(in),  dimension(:,:)   :: u_star, b_star
-real, intent(out), dimension(:,:,:) :: k_m, k_h
+real(kind=FMS_MO_KIND_), intent(in),  dimension(:,:,:) :: z
+real(kind=FMS_MO_KIND_), intent(in),  dimension(:,:)   :: u_star, b_star
+real(kind=FMS_MO_KIND_), intent(out), dimension(:,:,:) :: k_m, k_h
 
-integer            :: ni, nj, nk, ier
-real, parameter    :: ustar_min = 1.e-10
+integer                                                :: ni, nj, nk, ier
+real(kind=FMS_MO_KIND_), parameter                     :: ustar_min = 1.e-10
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov.inc:161]: should be
```fortran
real(kind=FMS_MO_KIND_), parameter                     :: ustar_min = real(1.d-10,FMS_MO_KIND_)
```

File: monin_obukhov/include/monin_obukhov.inc [hunk 2]
```diff
@@ -252,153 +94,174 @@ if(present(avail)) then
...
+real(kind=FMS_MO_KIND_), intent(in) , dimension(:) :: rich, z, z0, zt, zq
+logical, intent(in) , dimension(:)                 :: mask
+real(kind=FMS_MO_KIND_), intent(out), dimension(:) :: f_m, f_t, f_q
 
 
-real, parameter    :: error    = 1.e-04
-real, parameter    :: zeta_min = 1.e-06
-integer, parameter :: max_iter = 20
+real(kind=FMS_MO_KIND_), parameter                 :: error    = 1.e-04
+real(kind=FMS_MO_KIND_), parameter                 :: zeta_min = 1.e-06
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov.inc:187]: Same thing as above.

File: monin_obukhov/include/monin_obukhov.inc [hunk 3]
```diff
@@ -416,126 +279,133 @@ end do iter_loop
...
+integer, parameter                                  :: lkind = FMS_MO_KIND_
 
-stable   = mask .and. zeta >= 0.0
-unstable = mask .and. zeta <  0.0
+stable   = mask .and. zeta >= 0.0_lkind
+unstable = mask .and. zeta <  0.0_lkind
 
 where (unstable)
-  x     = (1 - 16.0*zeta  )**(-0.5)
+  x     = (1 - 16.0_lkind*zeta  )**(-0.5_lkind)
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov.inc:302]: The 1 has an implicit cast.  It should be `1.0_lkind`

File: monin_obukhov/include/monin_obukhov.inc [hunk 4]
```diff
@@ -416,126 +279,133 @@ end do iter_loop
...
+integer, parameter                                  :: lkind = FMS_MO_KIND_
 
-stable   = mask .and. zeta >= 0.0
-unstable = mask .and. zeta <  0.0
+stable   = mask .and. zeta >= 0.0_lkind
+unstable = mask .and. zeta <  0.0_lkind
 
 where (unstable)
-  phi_t = (1 - 16.0*zeta)**(-0.5)
+  phi_t = (1 - 16.0_lkind*zeta)**(-0.5_lkind)
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov.inc:344]: same as above

File: monin_obukhov/include/monin_obukhov.inc [hunk 5]
```diff
@@ -416,126 +279,133 @@ end do iter_loop
...
-    phi_t = 1.0 + zeta*(5.0 + b_stab*zeta)/(1.0 + zeta)
+    phi_t = 1.0_lkind + real(zeta, FMS_MO_KIND_)*(5.0_lkind + real(b_stab, FMS_MO_KIND_)&
+            *real(zeta,FMS_MO_KIND_))/(1.0_lkind + real(zeta, FMS_MO_KIND_))
   end where
 
 else if(stable_option == 2) then
 
   where (stable .and. zeta < zeta_trans)
-    phi_t = 1 + 5.0*zeta
+    phi_t = 1 + 5.0_lkind*zeta
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov.inc:357]: same as above

File: monin_obukhov/include/monin_obukhov.inc [hunk 6]
```diff
@@ -416,126 +279,133 @@ end do iter_loop
...
+unstable = mask .and. zeta <  0.0_lkind
 
 where(unstable)
 
-  x     = sqrt(1 - 16.0*zeta)
-  x_t   = sqrt(1 - 16.0*zeta_t)
-  x_q   = sqrt(1 - 16.0*zeta_q)
+  x     = sqrt(1 - 16.0_lkind*zeta)
+  x_t   = sqrt(1 - 16.0_lkind*zeta_t)
+  x_q   = sqrt(1 - 16.0_lkind*zeta_q)
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov.inc:392]: All of these 1s are implicitly cast as above.

File: monin_obukhov/include/monin_obukhov.inc [hunk 7]
```diff
@@ -545,64 +415,70 @@ else if (stable_option == 2) then
...
-unstable = mask .and. zeta <  0.0
+stable   = mask .and. zeta >= 0.0_lkind
+unstable = mask .and. zeta <  0.0_lkind
 
 where(unstable)
 
-  x     = sqrt(1 - 16.0*zeta)
-  x_0   = sqrt(1 - 16.0*zeta_0)
+  x      = sqrt(1 - 16.0_lkind*zeta)
+  x_0    = sqrt(1 - 16.0_lkind*zeta_0)
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov.inc:472]: More implicit casting

File: monin_obukhov/include/monin_obukhov_inter.inc [hunk 8]
```diff
@@ -22,44 +22,27 @@
...
-  integer, intent(in   )                        :: ni, nj, nk
-  real   , intent(in   ), dimension(ni, nj, nk) :: z
-  real   , intent(in   ), dimension(ni, nj)     :: u_star, b_star
-  real   , intent(  out), dimension(ni, nj, nk) :: k_m, k_h
-  integer, intent(  out)                        :: ier
-
-  real , dimension(ni, nj) :: phi_m, phi_h, zeta, uss
-  integer :: j, k
+  real(kind=FMS_MO_KIND_), intent(in)                         :: vonkarm
+  real(kind=FMS_MO_KIND_), intent(in)                         :: ustar_min !< = 1.e-10
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov_inter.inc:33]:
```fortran
real(kind=FMS_MO_KIND_), intent(in)                         :: ustar_min !< = 1.d-10
```

File: monin_obukhov/include/monin_obukhov_inter.inc [hunk 9]
```diff
@@ -312,147 +297,150 @@ pure subroutine monin_obukhov_solve_zeta(error, zeta_min, max_iter, small,  &
...
-  stable   = mask .and. zeta >= 0.0
-  unstable = mask .and. zeta <  0.0
+  stable   = mask .and. zeta >= 0.0_lkind
+  unstable = mask .and. zeta <  0.0_lkind
 
 !miz: modified to include new monin-obukhov option
   if (new_mo_option) then
      where (unstable)
-        phi_t = (1 - 16.0*zeta)**(-1./3.)
+        phi_t = (1 - 16.0_lkind*zeta)**(-1.0_lkind/3.0_lkind)
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov_inter.inc:330]: The `1` should be `1.0_lkind`

File: monin_obukhov/include/monin_obukhov_inter.inc [hunk 10]
```diff
@@ -312,147 +297,150 @@ pure subroutine monin_obukhov_solve_zeta(error, zeta_min, max_iter, small,  &
...
      end where
   else
   where (unstable)
-     phi_t = (1 - 16.0*zeta)**(-0.5)
+     phi_t = (1 - 16.0_lkind*zeta)**(-0.5_lkind)
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov_inter.inc:334]: Same as above

File: monin_obukhov/include/monin_obukhov_inter.inc [hunk 11]
```diff
@@ -312,147 +297,150 @@ pure subroutine monin_obukhov_solve_zeta(error, zeta_min, max_iter, small,  &
...
      end where
 
   else if(stable_option == 2) then
 
-     lambda = 1.0 + (5.0 - b_stab)*zeta_trans
+     lambda = 1.0_lkind + (5.0_lkind - b_stab)*zeta_trans
 
      where (stable .and. zeta < zeta_trans)
-        phi_t = 1 + 5.0*zeta
+        phi_t = 1 + 5.0_lkind*zeta
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov_inter.inc:350]: Same as above for `1`

File: monin_obukhov/include/monin_obukhov_inter.inc [hunk 12]
```diff
@@ -312,147 +297,150 @@ pure subroutine monin_obukhov_solve_zeta(error, zeta_min, max_iter, small,  &
...
 
-  stable   = mask .and. zeta >= 0.0
-  unstable = mask .and. zeta <  0.0
+  stable   = mask .and. zeta >= 0.0_lkind
+  unstable = mask .and. zeta <  0.0_lkind
 
   where (unstable)
-     x     = (1 - 16.0*zeta  )**(-0.5)
-     phi_m = sqrt(x)  ! phi_m = (1 - 16.0*zeta)**(-0.25)
+     x     = (1 - 16.0_lkind*zeta  )**(-0.5_lkind)
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov_inter.inc:385]: Same `1` as above

File: monin_obukhov/include/monin_obukhov_inter.inc [hunk 13]
```diff
@@ -312,147 +297,150 @@ pure subroutine monin_obukhov_solve_zeta(error, zeta_min, max_iter, small,  &
...
+     phi_m = sqrt(x) ! phi_m = (1 - 16.0*zeta)**(-0.25)
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov_inter.inc:386]: All of the numbers here need the kind added

File: monin_obukhov/include/monin_obukhov_inter.inc [hunk 14]
```diff
@@ -312,147 +297,150 @@ pure subroutine monin_obukhov_solve_zeta(error, zeta_min, max_iter, small,  &
...
      end where
 
   else if(stable_option == 2) then
 
-     lambda = 1.0 + (5.0 - b_stab)*zeta_trans
+     lambda = 1.0_lkind + (5.0_lkind - b_stab)*zeta_trans
 
      where (stable .and. zeta < zeta_trans)
-        phi_m = 1 + 5.0*zeta
+        phi_m = 1 + 5.0_lkind*zeta
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov_inter.inc:400]: same `1` needs a kind

File: monin_obukhov/include/monin_obukhov_inter.inc [hunk 15]
```diff
@@ -549,85 +538,88 @@ pure subroutine monin_obukhov_integral_m(stable_option, rich_crit, zeta_trans, &
...
-  psi_t = ln_z_zt - 1.5*log((x**2+x+1)/(x_t**2 + x_t + 1)) + s3*(atan((2*x+1)/s3) - atan((2*x_t + 1)/s3))
-  psi_q = ln_z_zq - 1.5*log((x**2+x+1)/(x_q**2 + x_q + 1)) + s3*(atan((2*x+1)/s3) - atan((2*x_q + 1)/s3))
+  x     = (1 - 16.0_lkind*zeta)**(1.0_lkind/3.0_lkind)
+  x_t   = (1 - 16.0_lkind*zeta_t)**(1.0_lkind/3.0_lkind)
+  x_q   = (1 - 16.0*zeta_q)**(1.0_lkind/3.0_lkind)
+
+  psi_t = ln_z_zt - 1.5_lkind*log((x**2+x+1)/(x_t**2 + x_t + 1)) + s3*(atan((2*x+1)/s3) - &
+               atan((2*x_t + 1)/s3))
+  psi_q = ln_z_zq - 1.5_lkind*log((x**2+x+1)/(x_q**2 + x_q + 1)) + s3*(atan((2*x+1)/s3) - &
+              atan((2*x_q + 1)/s3))
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov_inter.inc:611]: All of the integer `1` and `2`s need to be updated

File: monin_obukhov/include/monin_obukhov_inter.inc [hunk 16]
```diff
@@ -549,85 +538,88 @@ pure subroutine monin_obukhov_integral_m(stable_option, rich_crit, zeta_trans, &
...
 else
 
 where(unstable)
 
-  x     = sqrt(1 - 16.0*zeta)
-  x_t   = sqrt(1 - 16.0*zeta_t)
-  x_q   = sqrt(1 - 16.0*zeta_q)
+  x     = sqrt(1 - 16.0_lkind*zeta)
+  x_t   = sqrt(1 - 16.0_lkind*zeta_t)
+  x_q   = sqrt(1 - 16.0_lkind*zeta_q)
```
Comment (thomas-robinson) [inline monin_obukhov/include/monin_obukhov_inter.inc:619]: More `1`s

**Review by thomas-robinson — CHANGES_REQUESTED** [review]

**Review by rem1776 — CHANGES_REQUESTED** [review]

File: monin_obukhov/monin_obukhov.F90 [hunk 17]
```diff
@@ -48,29 +49,52 @@ module monin_obukhov_mod
...
-                      stable_mix_2d, stable_mix_3d
+    module procedure stable_mix_0d_r4, stable_mix_0d_r8
+    module procedure stable_mix_1d_r4, stable_mix_1d_r8
+    module procedure stable_mix_2d_r4, stable_mix_2d_r8
+    module procedure stable_mix_3d_r4, stable_mix_3d_r8
 end interface
 
+interface mo_integral
+    module procedure mo_integral_m_r4, mo_integral_m_r8
+    module procedure mo_integral_tq_r4, mo_integral_tq_r8
```
[the remark this answers is not in the document]
Comment (mcallic2) [reply monin_obukhov/monin_obukhov.F90:88]: What do you think would be best? The input/output arguments are a little different so whichever you think

**Comment by J-Lentz** [comment]
`m` in `STABLE_MIX_3D_` should get initialized prior to the loop where it's used.

**Comment by J-Lentz** [comment]
It seems that `MO_DRAG_1D_` in `monin_obukhov_mod` is passing its `avail` variable to `MONIN_OBUKHOV_DRAG_1D_` in `monin_obukhov_iter`. Problem is, `avail` is `optional` in the former subroutine, and non-optional in the latter. I'm not sure why it compiles without an error, but it seems to cause a segmentation fault if `MO_DRAG_1D_` is called without `avail`.

**Comment by J-Lentz** [comment]
The original version of `mo_drag_1d` in the main branch seems to solve this issue by passing `avail_dummy` rather than `avail` when `avail` is expected not to be present.

**Comment by J-Lentz** [comment]
I think all occurrences of `real(const, kind)` need to be changed to `const_kind`. Otherwise, the result can change depending on the presence or absence of an `-fdefault-real-*` flag.

**Comment by mlee03** [comment]
@mcallic2 could you also compile with -Wall and  -Wconversion-extra to catch all the type/kind mismatch?  There are several I think that have been missed.  After those fixes, I think the PR is good to be merged

File: CHANGELOG.md [hunk 18]
```diff
@@ -6,8 +6,6 @@ and this project uses `yyyy.rr[.pp]`, where `yyyy` is the year a patch is releas
 `rr` is a sequential release number (starting from `01`), and an optional two-digit
 sequential patch number (starting from `01`).
 
-<<<<<<< HEAD
```
Comment (mlee03) [inline CHANGELOG.md:9]: Thank for fixing this :)

File: monin_obukhov/include/monin_obukhov.inc [hunk 19]
```diff
@@ -16,190 +16,25 @@
...
 
-integer, parameter :: max_iter = 20
-real   , parameter :: error=1.e-04, zeta_min=1.e-06, small=1.e-04
+integer, parameter                                   :: max_iter = 20
+integer, parameter                                   :: lkind    = FMS_MO_KIND_
+real(kind=FMS_MO_KIND_), parameter                   :: error    = 1.0E-04_lkind, &
+                                                        zeta_min = 1.0E-06_lkind, &
+                                                        small    = 1.0E-04_lkind
 
-! #include "monin_obukhov_interfaces.h"
```
Comment (mlee03) [inline monin_obukhov/include/monin_obukhov.inc:202]: should we keep this commented out code?

Comment (mcallic2) [reply monin_obukhov/include/monin_obukhov.inc:202]: I don't recall if I was told to remove this or why I did this so I will follow your lead on whether or not to keep it in the code

Comment (mlee03) [reply monin_obukhov/include/monin_obukhov.inc:202]: Let's keep it in and let the future people remove it if they want to.

Comment (mcallic2) [reply monin_obukhov/include/monin_obukhov.inc:202]: yes ma'am

**Review by mlee03 — APPROVED** [review]

**Comment by mlee03** [comment]
@rem1776 @J-Lentz looks good?

**Comment by J-Lentz** [comment]
Yep, looks good to me.

**Review by rem1776 — APPROVED** [review]

**Pull request merged**